loopchain/consensus/runner.py

```diff
@@ -18,33 +30,83 @@ def __init__(self,
                  event_system: 'EventSystem',
                  data_factory: 'DataFactory',
                  vote_factory: 'VoteFactory',
-                 broadcast_scheduler: 'BroadcastScheduler'):
+                 broadcast_scheduler: 'BroadcastScheduler',
+                 block_manager):
         super().__init__(event_system.simulator)
         self.broadcast_scheduler = broadcast_scheduler
         self.event_system = event_system
         self.consensus = Consensus(self.event_system, node_id, data_factory, vote_factory)
+        self._block_manager: 'BlockManager' = block_manager
+        self._block_factory: 'BlockFactory' = data_factory
+        self._vote_factory: 'BlockVoteFactory' = vote_factory
 
     def start(self, event: InitializeEvent):
         self.event_system.start(blocking=False)
         self.event_system.simulator.raise_event(event)
 
-    async def _on_event_broadcast_data(self, block: 'Data'):
-        # call broadcast block
-        pass
+    async def _on_event_broadcast_data(self, event: BroadcastDataEvent):
+        target_reps_hash = ChannelProperty().crep_root_hash  # FIXME
+        self._block_manager._send_unconfirmed_block(
+            block_=event.data,
+            target_reps_hash=target_reps_hash,
+            round_=event.data.round_num
+        )
+
+    async def _on_event_broadcast_vote(self, event: BroadcastVoteEvent):
+        vote_dumped = self._vote_dumps(event.vote)
+        block_vote = loopchain_pb2.BlockVote(vote=vote_dumped, channel=ChannelProperty().name)
+
+        target_reps_hash = ChannelProperty().crep_root_hash  # FIXME
+
+        self.broadcast_scheduler.schedule_broadcast(
+            "VoteUnconfirmedBlock",
+            block_vote,
+            reps_hash=target_reps_hash
+        )
+
+    async def _on_round_end_event(self, round_end_event: RoundEndEvent):
+        utils.logger.notice(f"_on_round_end_event")
 
-    async def _on_event_broadcast_vote(self, vote: 'Vote'):
-        # call broadcast vote
-        pass
+        await self._write_block(round_end_event)
+        await self._round_start(round_end_event)
 
-    async def _on_event_receive_data(self, block: 'Data'):
-        await self.consensus.receive_data(block)
+    # FIXME: Temporary
+    async def _write_block(self, round_end_event):
+        utils.logger.notice(f"> EPOCH // ROUND ({round_end_event.epoch_num} // {round_end_event.round_num})")
+        utils.logger.notice(f"> Candidate id: {round_end_event.candidate_id}")
+        utils.logger.notice(f"> Commit id: {round_end_event.commit_id}")
 
-    async def _on_event_receive_vote(self, vote: 'Vote'):
-        await self.consensus.receive_vote(vote)
+        if round_end_event.is_success:
+            consensus_db_pool = self.consensus._data_pool  # FIXME
```

Comment:
How about using `CandidateBlocks` in `BlockManager` for getting a candidate block to commit? Accessing the data pool in `LFT` by the `loopchain` looks weird.

Reply:
> Accessing the data pool in LFT by the loopchain looks weird.
I totally agree.

LFT supplies ONLY `commit_id` to application, not the commited `Block`...
which means that Loopchain as application should access candidate blocks by using `commit_id`.

We have one of followings options:
- Use `CandidateBlocks` to get commited block.
- Use shared `DataPool` between app and LFT (currently, LFT has its own `DataPool`, initialized at the first moment of `Consensus`)
- Pass committed block with `round_end_event` (I'm afraid this makes lots of side effects or major changes.)

Any ideas?

Reply:
I think it's enough now, but We'll have to fix it later. Basically 'Use shared DataPool' is better.